<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Interpretable factor models of single-cell RNA-seq via variational autoencoders</title>
    <meta charset="utf-8" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/tile-view/tile-view.css" rel="stylesheet" />
    <script src="libs/tile-view/tile-view.js"></script>
    <link href="libs/panelset/panelset.css" rel="stylesheet" />
    <script src="libs/panelset/panelset.js"></script>
    <link rel="stylesheet" href="my-theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Interpretable factor models of single-cell RNA-seq via variational autoencoders
## V. Svensson, A. Gayoso, N. Yosef, and L. Pachter
### Kim Lab Journal Club
### 2021/03/24

---


# Introduction: TL; DR

.extraLarge[
- The author replaces the non-line decoder of scVI with a linear transformation.
- It gives interpretable loading for the latent variables while sacrificing the 
ability of reconstruction.
- Explore logit Normal vs. Normal distribution for the latent variable.
- Explore the effect of batch normalization during the training period.
]

---


# Table of Contents

.extraLarge[
1. Introduction
1. Review scVI
1. LDVAE modifies scVI
1. Trade-off: Reconstruction vs. Interpretability
1. Probabilistic PCA

]


&lt;!-- whereas typically autoencoder models are designed with the same network topology in the inference functions and the reconstruction functions, what we propose is a restricted reconstruction function that leads to an increase in reconstruction error. However, by virtue of being linear, our reconstruction function provides an interpretable link between gene programs and cellular molecular phenotypes --&gt;

???

---

# Introduction: VAE

.panelset[
.panel[.panel-name[Autoencoder]
&lt;img src="img/autoencoder.png" style="display: block; margin: auto;" /&gt;
]
.panel[.panel-name[Variational Autoencoder]
&lt;img src="img/vae.png" width="85%" style="display: block; margin: auto;" /&gt;
]
]
---

# Review scVI: Architecture

.panelset[
.panel[.panel-name[Encoder]
&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="img/model-1.png" alt="Variational Model" width="73%" /&gt;
&lt;p class="caption"&gt;Variational Model&lt;/p&gt;
&lt;/div&gt;
]
.panel[.panel-name[Decoder]
&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="img/model-2.png" alt="Generative Model" width="80%" /&gt;
&lt;p class="caption"&gt;Generative Model&lt;/p&gt;
&lt;/div&gt;
]
.panel[.panel-name[Loss Function]
Suppose:
- `\(x\)` is the cell by gene matrix `\((c, g)\)`. 
- `\(\hat{\mu}_{l}\)` and `\(\hat{\sigma}_{l}\)` are about size factors estimated from mini-batch.
- `\(\theta^g\)` is the gene-wise dispersion estimated from the dataset.
- `\(\mu\)` is the output from the decoder, as the rate of `\(\operatorname{NB}\)`.

Loss = Reconstruction Loss + KL Divergence Loss
- Reconstruction Loss
`$$-\log P(x| \operatorname{NB}(\mu, \theta^g))$$`
- KL Divergence Loss
`$$\operatorname{KL}(\mathcal{N}(\mu_z, \sigma_z), \mathcal{N}(0,1)) + \operatorname{KL}(\mathcal{N}(\mu_l, \sigma_l), \mathcal{N}(\hat{\mu}_{l}, \hat{\sigma}_{l}))$$`
]
]

&lt;!-- --- --&gt;

&lt;!-- # Review scVI: Randomness vs. Stability --&gt;

---

# LDVAE modifies scVI
.panelset[
.panel[.panel-name[LDVAE Encoder]
&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="img/model-3.png" alt="Generative Model" width="70%" /&gt;
&lt;p class="caption"&gt;Generative Model&lt;/p&gt;
&lt;/div&gt;
]
.panel[.panel-name[LDVAE Decoder]
&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="img/model-4.png" alt="Generative Model" width="80%" /&gt;
&lt;p class="caption"&gt;Generative Model&lt;/p&gt;
&lt;/div&gt;
]
.panel[.panel-name[Same Loss Function]
Suppose:
- `\(x\)` is the cell by gene matrix `\((c, g)\)`. 
- `\(\hat{\mu}_{l}\)` and `\(\hat{\sigma}_{l}\)` are about size factors estimated from mini-batch.
- `\(\theta^g\)` is the gene-wise dispersion estimated from the dataset.
- `\(\mu\)` is the output from the decoder, as the rate of `\(\operatorname{NB}\)`.

Loss = Reconstruction Loss + KL Divergence Loss
- Reconstruction Loss
`$$-\log P(x| \operatorname{NB}(\mu, \theta^g))$$`
- KL Divergence Loss
`$$\operatorname{KL}(\mathcal{N}(\mu_z, \sigma_z), \mathcal{N}(0,1)) + \operatorname{KL}(\mathcal{N}(\mu_l, \sigma_l), \mathcal{N}(\hat{\mu}_{l}, \hat{\sigma}_{l}))$$`
]
]

???

The only difference LDVAE makes to scVI is to replace the encoder with a linear transformation.

---

# Dataset
.pull-left[
- The first dataset (Pijuan-Sala et al., 2019) consists of 125 775 cells from 411 mouse embryos undergoing gastrulation measured using the commercial 10× Genomics platform and sequenced relatively deeply (11% non-zero values).
]
.pull-right[
- The second dataset (Cao et al., 2019) consists of 1 949 131 cells from 61 embryos in total using the sci-RNA-seq method at shallow sequencing (2% non-zero values). 
- This dataset is, to our knowledge, the largest scRNA-seq study published to date. 
]

???

To illustrate the scalability of our model, we fit a 10-dimensional LDVAE to the data which allows identification of cells similar to each other and for the determination of covarying genes
---

# Trade-off: Reconstruction
Using a normal latent distribution and batch norm has the smaller reconstruction error.

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="img/reconstruction.png" alt="Fig 1 (b) Comparison of reconstruction error on the Pijuan-Sala et al. (2019) data with VAE and the four variants of LDVAE after running 100 epochs." width="90%" /&gt;
&lt;p class="caption"&gt;Fig 1 (b) Comparison of reconstruction error on the Pijuan-Sala et al. (2019) data with VAE and the four variants of LDVAE after running 100 epochs.&lt;/p&gt;
&lt;/div&gt;

???
Question: Given the 

---

# Trade-off: Interpretability with `\(\mathcal{N}\)`
(c) Results from fitting a 20-dimensional LDVAE with Normal Distribution. Top genes indicated as vectors with names.

.panelset[
.panel[.panel-name[Fig-1c:row 1,2]
.pull-left[
- Left column: Density plots of the cells in representation space. 
- Right column: Scatter plots of gene loadings corresponding to the representation coordinates. 

]
.pull-right[
&lt;img src="img/fig1-c-1.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]

]
.panel[.panel-name[Fig-1c:row 4,5]
.pull-left[
The last row shows a pair of factors which discriminates three cell types annotated by Pijuan-Sala et al. (2019).
- red: erythroid
- blue: extraembryonic endoderm
- green: epiblast
]
.pull-right[
&lt;img src="img/fig1-c-2.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]

]
]

---

# Trade-off: Interpretability with `\(\operatorname{logit}\mathcal{N}\)`
Since a factor z is restricted to non-negative values, genes with negative weights `\(w\)` can only decrease in expression as a function of z⁠. This way cells using a particular regulatory program can more effectively be grouped to specific factors
.panelset[
.panel[.panel-name[Sup Fig-2:row 1,2]
.pull-left[
- Left column: Density plots of the cells in representation space. 
- Right column: Scatter plots of gene loadings corresponding to the representation coordinates. 

]
.pull-right[
&lt;img src="img/sup-fig-2-1.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]

]
.panel[.panel-name[Sup Fig-2:row 4,5]
.pull-left[
The last row shows a pair of factors which discriminates three cell types annotated by Pijuan-Sala et al. (2019).
- red: erythroid
- blue: extraembryonic endoderm
- green: epiblast
]
.pull-right[
&lt;img src="img/sup-fig-2-2.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]
]
]

---

# Trade-off: Interpretability with `\(\hat{Z}^T\hat{Z}\)`
Sup Fig 3 with Pijuan-sala et al 2019
.panelset[
.panel[.panel-name[LDVAE Batch Norm]
Simplicial structure: one factor is always linearly dependent on the others.

.pull-left[
&lt;img src="img/sup-fig-3-1.png" width="100%" style="display: block; margin: auto auto auto 0;" /&gt;
]
.pull-right[
&lt;img src="img/sup-fig-3-2.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]

]
.panel[.panel-name[LDVAE no Batch Norm]
Simplicial structure: one factor is always linearly dependent on the others.

.pull-left[
&lt;img src="img/sup-fig-3-3.png" width="100%" style="display: block; margin: auto auto auto 0;" /&gt;
]
.pull-right[
&lt;img src="img/sup-fig-3-4.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]
]
.panel[.panel-name[LDVAE vs. scVI]
LDVAE learns representations with fewer covarying factors.
.pull-left[
&lt;img src="img/sup-fig-3-2.png" width="100%" style="display: block; margin: auto auto auto 0;" /&gt;
]
.pull-right[
&lt;img src="img/sup-fig-3-5.png" width="100%" style="display: block; margin: auto 0 auto auto;" /&gt;
]
]
]
---

# Discussion: `\(\hat{Z}^T\hat{Z}\)` Plot

I cannot observe the Simplicial structure...

&lt;img src="img/sup-fig-3-mock.png" width="85%" style="display: block; margin: auto;" /&gt;

---

# Discussion: PCA

In the introduction section, the author mentioned that 
&gt; PCA models data as arising from a continuous multivariate Gaussian distribution, and thus optimizes a Gaussian likelihood. 

Also, 
&gt; This model assumption is at odds with the count data measured in single-cell RNA-seq, and leads to interpretation problems (Hicks et al., 2018). 

I believe everyone agree with the author that PCA is not the best tools for 
scRNA-seq data. But the author's reason is about the assumption of 
Probabilistic PCA, which is different from the standard PCA in Seurat and Scanpy.

---

## Standar PCA

Use orthogonal projection of the data onto a principle subspace such that
- Maximum the variance of the projected data.
- Minimize the squared distance between the data points and the projections.

The two definitions are equivalent and can be solved by SVD.

## Probabilistic PCA
`$$\boldsymbol{z} \sim N(0, \boldsymbol{I})$$`
`$$\boldsymbol{x} \sim N(\boldsymbol{Wz} + \boldsymbol{\mu}, \sigma^2 \boldsymbol{I})$$`
`$$\boldsymbol{W}_{ML} = \boldsymbol{U}_M(\boldsymbol{L}_M - \sigma^2 \boldsymbol{I})^{1/2}\boldsymbol{R}$$`
- `\(\boldsymbol{U}_M\)` is a combination of eigenvectors of the data covariance matrix.
- `\(\boldsymbol{L}_M\)` is a diagonal matrix with eigenvalues `\(\lambda_i\)`.
- `\(\boldsymbol{R}\)` is an arbitrary orthogonal matrix.
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
